### ISSUE TITLE
conda skeleton picked oldest version instead of newest

### ISSUE THREAD

**Issue by ijstokes** [conda/conda-build#974]
_From @ijstokes on May 25, 2016 3:47_

Oldest version was 2.0.8, newest was 2.8.0.  `conda skeleton` picked 2.0.8.  Weird!

```
$ conda skeleton pypi rpy2
Warning, the following versions were found for rpy2
2.0.8
2.1.0
2.1.1
2.1.2
2.1.3
2.1.4
2.1.5
2.1.6
2.1.7
2.1.8
2.1.9
2.2.0
2.2.1
2.2.2
2.2.3
2.2.4
2.2.5
2.2.6
2.2.7
2.3.0alpha1
2.3.0beta1
2.3.0
2.3.1
2.3.2
2.3.3
2.3.4
2.3.5
2.3.6
2.3.7
2.3.8
2.3.9
2.3.10
2.4.0
2.4.1
2.4.2
2.4.3
2.4.4
2.5.0
2.5.1
2.5.2
2.5.3
2.5.4
2.5.5
2.5.6
2.6.0
2.6.1.dev0
2.6.1
2.6.2
2.6.3
2.7.0
2.7.1
2.7.2
2.7.3
2.7.4
2.7.5
2.7.6
2.7.7
2.7.8
2.7.9
2.8.0
Using 2.0.8
```

This is with conda 4.1.0rc3

_Copied from original issue: conda/conda#2549_

**Comment by ijstokes** [comment]
_From @msarahan on May 25, 2016 4:8_

I have seen things like this. I think it has to do with source releases vs wheel releases. Often times later versions are wheel only. 

This is really a conda build issue. Feel free to move it there or refile there.

**Comment by ijstokes** [comment]
So further to @msarahan 's comment, the weird thing is that `conda skeleton` can still see the 2.8.0 version (it is in the list), and if I use the `--version` option I can specify that 2.8.0 should be used and that works.  So this definitely seems like a bug to me.

**Comment by mcg1969** [comment]
@msarahan does `conda skeleton` utilize conda's `VersionOrder` logic at all?

**Comment by jakirkham** [comment]
Also, seeing this problem BTW.

**Comment by jakirkham** [comment]
Should add that I'm using standard (non-dev) releases.

```
conda 4.0.6
conda-build 1.20.3
```

**Comment by bmcguirk** [comment]
@mcg1969 - [It appears](https://github.com/conda/conda-build/blob/master/conda_build/pypi.py) that it does: 

``` python
def version_compare(args, package, versions):
    if not versions:
        # PyPI is case sensitive, this will pass control
        # to a method in main() to take care of that.
        return

    nv = normalized_version
```

Normalized_version is pulled here: 
`from conda.resolve import normalized_version`

Which is [pulling](https://github.com/conda/conda/blob/master/conda/resolve.py) from version:
```
from .version import VersionSpec, normalized_version
```

And here's the code in version: 

```python
def normalized_version(version):
    return VersionOrder(version)
```

and: 

```python
class VersionOrder(object):
    '''
    This class implements an order relation between version strings.
    Version strings can contain the usual alphanumeric characters
    (A-Za-z0-9), separated into components by dots and underscores. Empty
    segments (i.e. two consecutive dots, a leading/trailing underscore)
    are not permitted. An optional epoch number - an integer
    followed by '!' - can preceed the actual version string
    (this is useful to indicate a change in the versioning
    scheme itself). Version comparison is case-insensitive.
    Conda supports six types of version strings:
    * Release versions contain only integers, e.g. '1.0', '2.3.5'.
    * Pre-release versions use additional letters such as 'a' or 'rc',
      for example '1.0a1', '1.2.beta3', '2.3.5rc3'.
    * Development versions are indicated by the string 'dev',
      for example '1.0dev42', '2.3.5.dev12'.
    * Post-release versions are indicated by the string 'post',
      for example '1.0post1', '2.3.5.post2'.
    * Tagged versions have a suffix that specifies a particular
      property of interest, e.g. '1.1.parallel'. Tags can be added
      to any of the preceding four types. As far as sorting is concerned,
      tags are treated like strings in pre-release versions.
    * An optional local version string separated by '+' can be appended
      to the main (upstream) version string. It is only considered
      in comparisons when the main versions are equal, but otherwise
      handled in exactly the same manner.
    To obtain a predictable version ordering, it is crucial to keep the
    version number scheme of a given package consistent over time.
    Specifically,
    * version strings should always have the same number of components
      (except for an optional tag suffix or local version string),
    * letters/strings indicating non-release versions should always
      occur at the same position.
    Before comparison, version strings are parsed as follows:
    * They are first split into epoch, version number, and local version
      number at '!' and '+' respectively. If there is no '!', the epoch is
      set to 0. If there is no '+', the local version is empty.
    * The version part is then split into components at '.' and '_'.
    * Each component is split again into runs of numerals and non-numerals
    * Subcomponents containing only numerals are converted to integers.
    * Strings are converted to lower case, with special treatment for 'dev'
      and 'post'.
    * When a component starts with a letter, the fillvalue 0 is inserted
      to keep numbers and strings in phase, resulting in '1.1.a1' == 1.1.0a1'.
    * The same is repeated for the local version part.
    Examples:
        1.2g.beta15.rc  =>  [[0], [1], [2, 'g'], [0, 'beta', 15], [0, 'rc']]
        1!2.15.1_ALPHA  =>  [[1], [2], [15], [1, '_alpha']]
    The resulting lists are compared lexicographically, where the following
    rules are applied to each pair of corresponding subcomponents:
    * integers are compared numerically
    * strings are compared lexicographically, case-insensitive
    * strings are smaller than integers, except
    * 'dev' versions are smaller than all corresponding versions of other types
    * 'post' versions are greater than all corresponding versions of other types
    * if a subcomponent has no correspondent, the missing correspondent is
      treated as integer 0 to ensure '1.1' == '1.1.0'.
    The resulting order is:
           0.4
         < 0.4.0
         < 0.4.1.rc
        == 0.4.1.RC   # case-insensitive comparison
         < 0.4.1
         < 0.5a1
         < 0.5b3
         < 0.5C1      # case-insensitive comparison
         < 0.5
         < 0.9.6
         < 0.960923
         < 1.0
         < 1.1dev1    # special case 'dev'
         < 1.1a1
         < 1.1.0dev1  # special case 'dev'
        == 1.1.dev1   # 0 is inserted before string
         < 1.1.a1
         < 1.1.0rc1
         < 1.1.0
        == 1.1
         < 1.1.0post1 # special case 'post'
        == 1.1.post1  # 0 is inserted before string
         < 1.1post1   # special case 'post'
         < 1996.07.12
         < 1!0.4.1    # epoch increased
         < 1!3.1.1.6
         < 2!0.4.1    # epoch increased again
    Some packages (most notably openssl) have incompatible version conventions.
    In particular, openssl interprets letters as version counters rather than
    pre-release identifiers. For openssl, the relation
      1.0.1 < 1.0.1a   =>   True   # for openssl
    holds, whereas conda packages use the opposite ordering. You can work-around
    this problem by appending a dash to plain version numbers:
      1.0.1a  =>  1.0.1post.a      # ensure correct ordering for openssl
    '''
    def __init__(self, version):
        # when fillvalue ==  0  =>  1.1 == 1.1.0
        # when fillvalue == -1  =>  1.1  < 1.1.0
        self.fillvalue = 0

        message = "Malformed version string '%s': " % version
        # version comparison is case-insensitive
        version = version.strip().rstrip().lower()
        # basic validity checks
        if version == '':
            raise ValueError("Empty version string.")
        invalid = not version_check_re.match(version)
        if invalid and '-' in version and '_' not in version:
            # Allow for dashes as long as there are no underscores
            # as well, by converting the former to the latter.
            version = version.replace('-', '_')
            invalid = not version_check_re.match(version)
        if invalid:
            raise ValueError(message + "invalid character(s).")
        self.norm_version = version

        # find epoch
        version = version.split('!')
        if len(version) == 1:
            # epoch not given => set it to '0'
            epoch = ['0']
        elif len(version) == 2:
            # epoch given, must be an integer
            if not version[0].isdigit():
                raise ValueError(message + "epoch must be an integer.")
            epoch = [version[0]]
        else:
            raise ValueError(message + "duplicated epoch separator '!'.")

        # find local version string
        version = version[-1].split('+')
        if len(version) == 1:
            # no local version
            self.local = []
        elif len(version) == 2:
            # local version given
            self.local = version[1].replace('_', '.').split('.')
        else:
            raise ValueError(message + "duplicated local version separator '+'.")

        # split version
        self.version = epoch + version[0].replace('_', '.').split('.')

        # split components into runs of numerals and non-numerals,
        # convert numerals to int, handle special strings
        for v in (self.version, self.local):
            for k in range(len(v)):
                c = version_split_re.findall(v[k])
                if not c:
                    raise ValueError(message + "empty version component.")
                for j in range(len(c)):
                    if c[j].isdigit():
                        c[j] = int(c[j])
                    elif c[j] == 'post':
                        # ensure number < 'post' == infinity
                        c[j] = float('inf')
                    elif c[j] == 'dev':
                        # ensure '*' < 'DEV' < '_' < 'a' < number
                        # by upper-casing (all other strings are lower case)
                        c[j] = 'DEV'
                if v[k][0].isdigit():
                    v[k] = c
                else:
                    # components shall start with a number to keep numbers and
                    # strings in phase => prepend fillvalue
                    v[k] = [self.fillvalue] + c

    def __str__(self):
        return self.norm_version

    def _eq(self, t1, t2):
        for v1, v2 in zip_longest(t1, t2, fillvalue=[]):
            for c1, c2 in zip_longest(v1, v2, fillvalue=self.fillvalue):
                if c1 != c2:
                    return False
        return True

    def __eq__(self, other):
        return (self._eq(self.version, other.version) and
                self._eq(self.local, other.local))

    def startswith(self, other):
        # Tests if the version lists match up to the last element in "other".
        if other.local:
            if not self._eq(self.version, other.version):
                return False
            t1 = self.local
            t2 = other.local
        elif other.version:
            t1 = self.version
            t2 = other.version
        else:
            return True
        nt = len(t2) - 1
        if not self._eq(t1[:nt], t2[:nt]):
            return False
        v1 = [] if len(t1) <= nt else t1[nt]
        v2 = t2[nt]
        nt = len(v2) - 1
        if not self._eq([v1[:nt]], [v2[:nt]]):
            return False
        c1 = self.fillvalue if len(v1) <= nt else v1[nt]
        c2 = v2[nt]
        if isinstance(c2, string_types):
            return isinstance(c1, string_types) and c1.startswith(c2)
        return c1 == c2

    def __ne__(self, other):
        return not (self == other)

    def __lt__(self, other):
        for t1, t2 in zip([self.version, self.local], [other.version, other.local]):
            for v1, v2 in zip_longest(t1, t2, fillvalue=[]):
                for c1, c2 in zip_longest(v1, v2, fillvalue=self.fillvalue):
                    if c1 == c2:
                        continue
                    elif isinstance(c1, string_types):
                        if not isinstance(c2, string_types):
                            # str < int
                            return True
                    elif isinstance(c2, string_types):
                            # not (int < str)
                            return False
                    # c1 and c2 have the same type
                    return c1 < c2
        # self == other
        return False

    def __gt__(self, other):
        return other < self

    def __le__(self, other):
        return not (other < self)

    def __ge__(self, other):
        return not (self < other)


# This RE matches the operators '==', '!=', '<=', '>=', '<', '>'
# followed by a version string. It rejects expressions like
# '<= 1.2' (space after operator), '<>1.2' (unknown operator),
# and '<=!1.2' (nonsensical operator).
version_relation_re = re.compile(r'(==|!=|<=|>=|<|>)(?![=<>!])(\S+)$')
opdict = {'==': op.__eq__, '!=': op.__ne__, '<=': op.__le__,
          '>=': op.__ge__, '<': op.__lt__, '>': op.__gt__}
```


And now that I've written all this, it appears you know more about this than I do, as you have commits directly related to this code. Is it possible something broke when modifying the wildcard processing in your most recent one?

**Comment by mcg1969** [comment]
Oh yes, it is _always_ possible that I broke something ;-) Just ask my fellow `conda` developers. 

But it would have to be a pretty weird combination of things that causes this, since `VersionOrder` does have pretty strong unit tests and of course a serious break in version ordering would manifest throughout `conda`'s unit tests, too.

Furthermore, the only recent change I made to VersionOrder logic itself was the addition of wildcard processing, and that isn't being engaged at all here, as far as I can tell.

**Comment by bmcguirk** [comment]
@mcg1969 - Quick thought from a dilettante: is there an opportunity to put in a test case here? Like for an essential module (like Conda?) you can assume that, at the very least, the version that the algorithm would choose should be higher than the current version? Maybe not the most elegant, just thinking about this. Not having it work is cramping my style. :)

**Comment by mcg1969** [comment]
@bmcguirk if @msarahan needs my help I'm happy to offer it, but I'm going to have to wait on his lead here.

**Comment by msarahan** [comment]
@mcg1969 - have at it if you have time.  I'm a little overwhelmed with a few of the other issues.

**Comment by mcg1969** [comment]
Well, I'm sorry I wasn't clear---I don't have the time, particularly since it would involve getting up to speed on areas of conda-build internals I'm not familiar with. But the minute you or anyone can point me to a section of the code and say "the version sorting is happening _here_, and it's not working", I'm on it.

**Comment by bmcguirk** [comment]
@mcg1969 @msarahan - Haha, sorry guys. I assumed @mcg1969 was actively developing because of recent commits, that's all. I'm, again, relatively new to Python generally, so I'm kind of useless here. I'll keep tabs on this issue though. And thanks to both of you for such an awesome tool.

**Comment by ijstokes** [comment]
Again:

```
$ conda skeleton pypi pydoop
Warning, the following versions were found for pydoop
0.9.0
0.9.1
0.10.0_rc1
0.10.0
0.11.1
0.12.0
1.0.0-rc1
1.0.0-rc2
1.0.0
1.1.0
1.2.0
Using 0.9.0
Use --version to specify a different version.
``

This is with:

```
$ conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.5
    conda-env version : 2.5.1
  conda-build version : 1.21.3
       python version : 2.7.12.final.0
     requests version : 2.9.1
     root environment : /opt/anaconda  (read only)
  default environment : /opt/anaconda
     envs directories : /home/ijstokes/.conda/envs
                        /home/ijstokes/envs
                        /opt/anaconda/envs
        package cache : /home/ijstokes/.conda/envs/.pkgs
                        /home/ijstokes/envs/.pkgs
                        /opt/anaconda/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : None
         offline mode : False
    is foreign system : False
```

**Comment by msarahan** [comment]
Thanks for the ping on this.  I hope to get to it soon.

**Comment by msarahan** [comment]
Fixed by #1092
